Lib/http/client.py

```diff
@@ -943,21 +944,15 @@ def _tunnel(self):
         response = self.response_class(self.sock, method=self._method)
         (version, code, message) = response._read_status()
 
+        self._proxy_response_headers = parse_headers(response.fp)
+
+        if self.debuglevel > 0:
+            for hdr, val in self._proxy_response_headers.items():
+                print("header:", hdr + ":", val)
+
         if code != http.HTTPStatus.OK:
             self.close()
             raise OSError(f"Tunnel connection failed: {code} {message.strip()}")
-        while True:
-            line = response.fp.readline(_MAXLINE + 1)
-            if len(line) > _MAXLINE:
-                raise LineTooLong("header line")
-            if not line:
```

Comment:
This condition `if not line:` isn't captured in `parse_headers` call.
Additional check of _MAXHEADERS is verified. 
So, this isn't a strict 1:1 replacement.

Reply:
We can add the `if not line` condition in _reader_headers method for equivalence.

However, given that _tunnel method already prints headers in line 960 in `print('header:', line.decode())`, I am trying to understand, what additional benefit this patch brings

The discussion in https://github.com/python/cpython/issues/69152 seems  adding additional states to the _tunnel method, and adding debug response to those states.

Reply:
@orsenthil, thanks for the review!
Regarding condition `if not line`. It seems to me that this is now a useless part of the code. Because the only variant of the line value in which this condition will be true, if I'm not mistaken, is `line = b"` (because the method `readline` can't return `None` or just an empty string). And for the equality of line to the value of b'', we check a little below.

I looked at the history, these lines were [added](https://github.com/python/cpython/commit/b12771ab3048bb61b3dc09fd21198613f43fa030) when there was no equality check for the empty binary object below. That is, at some point they made sense, but after this [commit](https://github.com/python/cpython/commit/7e70a5c16918386ced63a36efb79257b75fa7500), they ceased to be necessary. 

But we can add this check to the _read_headers method as a precaution. Do you think it should be done?

Reply:
2. As for the benefits of the changes being made. Firstly, headers are needed in the situations described in the comment https://github.com/python/cpython/issues/69152#issuecomment-1093691886. And secondly, now headers gets into the debug-logs only if we were able to go beyond this line: https://github.com/python/cpython/blob/main/Lib/http/client.py#L948. 
But if we did not pass the authorization headers we need, then we will get an OSError without information about what happened, we will not be able to find out what authentication data is required. With my PR, I would like to take the first step towards solving the problems described here https://github.com/psf/requests/issues/2526, https://github.com/psf/requests/issues/1582, https://github.com/requests/requests-kerberos/issues/83

Maybe then it would be possible to start returning not OSError, but some custom error that can be intercepted in the library code (urllib3, requests) and automatically prepare data for authentication based on the information contained in _proxy_response_headers. Similar to how it is done in https://github.com/requests/requests-kerberos/pull/149/files.

Maybe you know ways to solve these problems in a different way, I will be glad of any ideas and suggestions.
How @vadmium's proposal will solve the problem of the lack of any information about what went wrong when creating the tunnel, I did not understand.